### ISSUE TITLE
assignment 3

### ISSUE THREAD

**Issue by UMAGAYATHRI74** [UMAGAYATHRI74/data-analyst#1]
Assignment 3

https://www.online-python.com/h2dMyEnTLx

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.cluster import KMeans
from sklearn.metrics import davies_bouldin_score, silhouette_score
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

try:
    customers = pd.read_csv('Customers.csv')
    products = pd.read_csv('Products.csv')
    transactions = pd.read_csv('Transactions.csv')
except FileNotFoundError:
    print("Error: One or more CSV files not found. Please ensure they are in the same directory.")
    exit()

customers['SignupDate'] = pd.to_datetime(customers['SignupDate'])
transactions['TransactionDate'] = pd.to_datetime(transactions['TransactionDate'])

products['Price'].fillna(products['Price'].mean(), inplace=True)
transactions.dropna(subset=['CustomerID', 'ProductID'], inplace=True)

transactions = transactions.merge(customers[['CustomerID', 'Region']], on='CustomerID', how='left')
transactions = transactions.merge(products[['ProductID', 'ProductName', 'Category', 'Price']], on='ProductID', how='left')

customer_profiles = transactions.groupby('CustomerID').agg(
    total_transactions=('TransactionID', 'count'),
    total_spent=('TotalValue', 'sum'),
    avg_order_value=('TotalValue', 'mean'),
    unique_products=('ProductID', 'nunique'),
    electronics_purchases=('Category', lambda x: (x == 'Electronics').sum()),
    clothing_purchases=('Category', lambda x: (x == 'Clothing').sum()),
    home_purchases=('Category', lambda x: (x == 'Home').sum()),
    books_purchases=('Category', lambda x: (x == 'Books').sum()),
    region=('Region', 'first')
).reset_index()

customer_profiles = customer_profiles.merge(customers[['CustomerID', 'Region', 'SignupDate']], on='CustomerID', how='left')
customer_profiles['days_since_signup'] = (pd.to_datetime('today') - customer_profiles['SignupDate']).dt.days

customer_profiles = pd.get_dummies(customer_profiles, columns=['region'], prefix='region')
customer_profiles.drop('SignupDate', axis=1, inplace=True)

numerical_features = ['total_transactions', 'total_spent', 'avg_order_value', 'unique_products', 'electronics_purchases', 'clothing_purchases', 'home_purchases', 'books_purchases', 'days_since_signup']
scaler = StandardScaler()
customer_profiles[numerical_features] = scaler.fit_transform(customer_profiles[numerical_features])

best_db_index = -1
best_silhouette_score = -1
best_n_clusters = 0
best_labels = None
best_cluster_centers = None

for n_clusters in range(2, 11):
    kmeans = KMeans(n_clusters=n_clusters, random_state=42, n_init=10)
    labels = kmeans.fit_predict(customer_profiles.drop('CustomerID', axis=1))
    db_index = davies_bouldin_score(customer_profiles.drop('CustomerID', axis=1), labels)
    silhouette = silhouette_score(customer_profiles.drop('CustomerID', axis=1), labels)

    if db_index < best_db_index or best_db_index == -1:  # Lower DB index is better
        best_db_index = db_index
        best_silhouette_score = silhouette
        best_n_clusters = n_clusters
        best_labels = labels
        best_cluster_centers = kmeans.cluster_centers_

customer_profiles['cluster'] = best_labels

print(f"Best number of clusters: {best_n_clusters}")
print(f"DB Index: {best_db_index}")
print(f"Silhouette Score: {best_silhouette_score}")

pca = PCA(n_components=2)
pca_data = pca.fit_transform(customer_profiles.drop(['CustomerID', 'cluster'], axis=1))
customer_profiles['pca_1'] = pca_data[:, 0]
customer_profiles['pca_2'] = pca_data[:, 1]

plt.figure(figsize=(10, 8))
for cluster in range(best_n_clusters):
    cluster_data = customer_profiles[customer_profiles['cluster'] == cluster]
    plt.scatter(cluster_data['pca_1'], cluster_data['pca_2'], label=f'Cluster {cluster}')

plt.title('Customer Segmentation')
plt.xlabel('PCA Component 1')
plt.ylabel('PCA Component 2')
plt.legend()
plt.show()

for cluster in range(best_n_clusters):
    print(f"\nCluster {cluster} Characteristics:")
    cluster_data = customer_profiles[customer_profiles['cluster'] == cluster]
    print(cluster_data[numerical_features].describe())
    print(cluster_data['region_Europe'].value_counts())
